File: backend/app/schemas/artifact_custody.py
"""Artifact Versioning + Chain-of-Custody — Pydantic schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

# ---------- ArtifactVersion ----------


class ArtifactVersionCreate(BaseModel):
    artifact_type: str
    artifact_id: UUID
    content_hash: str | None = None


class ArtifactVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    artifact_type: str
    artifact_id: UUID
    version_number: int
    content_hash: str | None = None
    status: str
    superseded_by_id: UUID | None = None
    created_by_user_id: UUID | None = None
    created_at: datetime
    archived_at: datetime | None = None
    archive_reason: str | None = None


# ---------- CustodyEvent ----------


class CustodyEventCreate(BaseModel):
    artifact_version_id: UUID
    event_type: str
    actor_type: str = "system"
    actor_id: UUID | None = None
    actor_name: str | None = None
    recipient_org_id: UUID | None = None
    details: dict | None = None
    occurred_at: datetime | None = None


class CustodyEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    artifact_version_id: UUID
    event_type: str
    actor_type: str
    actor_id: UUID | None = None
    actor_name: str | None = None
    recipient_org_id: UUID | None = None
    details: dict | None = None
    occurred_at: datetime
    created_at: datetime


# ---------- Composite views ----------


class CustodyChainRead(BaseModel):
    artifact_type: str
    artifact_id: UUID
    current_version: ArtifactVersionRead | None = None
    versions: list[ArtifactVersionRead] = []
    events: list[CustodyEventRead] = []


class ArchivePostureRead(BaseModel):
    building_id: UUID
    total_artifacts: int = 0
    total_versions: int = 0
    superseded_count: int = 0
    archived_count: int = 0
    withdrawn_count: int = 0
    current_count: int = 0
    last_custody_event: CustodyEventRead | None = None


class ArchiveRequest(BaseModel):
    reason: str

File: frontend/src/api/artifactCustody.ts
import { apiClient } from '@/api/client';

export interface ArtifactVersion {
  id: string;
  artifact_type: string;
  artifact_id: string;
  version_number: number;
  content_hash: string | null;
  status: 'current' | 'superseded' | 'archived' | 'withdrawn';
  superseded_by_id: string | null;
  created_by_user_id: string | null;
  created_at: string;
  archived_at: string | null;
  archive_reason: string | null;
}

export interface CustodyEvent {
  id: string;
  artifact_version_id: string;
  event_type: string;
  actor_type: string;
  actor_id: string | null;
  actor_name: string | null;
  recipient_org_id: string | null;
  details: Record<string, unknown> | null;
  occurred_at: string;
  created_at: string;
}

export interface CustodyChain {
  artifact_type: string;
  artifact_id: string;
  current_version: ArtifactVersion | null;
  versions: ArtifactVersion[];
  events: CustodyEvent[];
}

export interface ArchivePosture {
  building_id: string;
  total_artifacts: number;
  total_versions: number;
  superseded_count: number;
  archived_count: number;
  withdrawn_count: number;
  current_count: number;
  last_custody_event: CustodyEvent | null;
}

export const artifactCustodyApi = {
  getArchivePosture: async (buildingId: string): Promise<ArchivePosture> => {
    const { data } = await apiClient.get(`/buildings/${buildingId}/archive-posture`);
    return data;
  },

  getChain: async (artifactType: string, artifactId: string): Promise<CustodyChain> => {
    const { data } = await apiClient.get(`/artifacts/${artifactType}/${artifactId}/chain`);
    return data;
  },

  getCurrentVersion: async (artifactType: string, artifactId: string): Promise<ArtifactVersion | null> => {
    const { data } = await apiClient.get(`/artifacts/${artifactType}/${artifactId}/current`);
    return data;
  },

  getVersionEvents: async (versionId: string): Promise<CustodyEvent[]> => {
    const { data } = await apiClient.get(`/artifacts/versions/${versionId}/events`);
    return data;
  },

  archiveVersion: async (versionId: string, reason: string): Promise<ArtifactVersion> => {
    const { data } = await apiClient.post(`/artifacts/versions/${versionId}/archive`, { reason });
    return data;
  },

  withdrawVersion: async (versionId: string): Promise<ArtifactVersion> => {
    const { data } = await apiClient.post(`/artifacts/versions/${versionId}/withdraw`);
    return data;
  },
};
